### ISSUE TITLE
Add Token Rotation Code

### ISSUE THREAD

**Issue by kevinhughes27** [Shopify/shopify_app#386]
Token rotation is something that ShopifyApp can help app developers with. During the rotation the app can use the new and old keys- the code will need to be updated in a way that allows for this. The other main component of the rotation is a job that makes a request for each shop to update the token.

Here is a ruby class I've used to do the rotation request in the past

```ruby
require 'net/http'
require 'uri'

module ShopifyAPI
  class RotateAccessToken
    attr_reader :api_key, :secret, :refresh_token

    def initialize(api_key:, secret:, refresh_token:)
      @api_key = api_key
      @secret = secret
      @refresh_token = refresh_token
    end

    def for_shop(shop_domain)
      parse_response(request(shop_domain))
    rescue => e
      Rails.logger.error("Token rotation failed with exception #{e}")
      :no_token
    end

    private

    def request(shop_domain)
      uri = URI.parse("https://#{shop_domain}/admin/oauth/access_token.json")
      Net::HTTP.post_form(uri, post_data)
    end

    def parse_response(response)
      if response.code.to_i == 200
        JSON.parse(response.body)['access_token']
      else
        Rails.logger.error("Token rotation failed with HTTP response code #{response.code}")
        :no_token
      end
    end

    def post_data
      @post_data ||= {
        client_id: api_key,
        client_secret: secret,
        refresh_token: refresh_token
      }
    end
  end
end
```

This would be a great first contribution :)

**Comment by jaydenwindle** [comment]
@kevinhughes27 I'd love to take a crack at this! Do you think a custom rake task would work well for running the rotation job, or did you have another method in mind?

> During the rotation the app can use the new and old keys- the code will need to be updated in a way that allows for this.

Could you clarify that a bit?

**Comment by kevinhughes27** [comment]
I looked through the code on one of our apps where we rotated tokens and it looks like the only place we did anything special was when receiving webhooks. We updated [this code](https://github.com/Shopify/shopify_app/blob/master/lib/shopify_app/webhook_verification.rb#L12) to check with both secrets since there is a bit of a race situation there.

I can't think of anywhere else we need to do this so I think thats it.

A rake task would be a good way to start the job but it might be better to implement a class that does the work so its easier to test and just have the rake task call this code. (I've found testing rake tasks to be less than awesome but maybe I just did it wrong)
